google/cloud/documentai_toolbox/wrappers/document_revision.py

```diff
@@ -0,0 +1,502 @@
+        ):
+            entities_array.append(Entity(documentai_entity=e))
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                }
+            )
+        elif e.provenance.type_ == _OP_TYPE.REMOVE:
+            entity = entities_array.pop(int(e.id) - 1)
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": entity.type_,
+                    "original_text": entity.mention_text,
+                }
+            )
+            del entity
+        elif e.provenance.type_ == _OP_TYPE.REPLACE:
+            history.append(
+                {
+                    "object": "Entity",
+                    "entity_provenance_type": _OP_TYPE(e.provenance.type_).name,
+                    "original_entity": e,
+                    "original_type": e.type_,
+                    "original_text": e.mention_text,
+                    "replace_type": e.type_,
+                    "replace_text": e.mention_text,
+                }
+            )
+            entities_array[int(e.id) - 1] = e
+    return entities_array, history
+
+
+def _get_level(current_revision: "DocumentWithRevisions") -> int:
+    r"""Returns the level of the current revision in the revision tree.
+
+    Args:
+        current_revision (DocumentWithRevisions):
+            The current revision.
+
+    Returns:
+        int:
+            The level of the current revision in the revision tree.
+    """
+    try:
+        level = 0
+        p = current_revision.parent
+        if p.revision_id:
+            while p:
+                p = p.parent
+                level += 1
+        return level
+    except Exception:
+        return 0
+
+
+def _print_child_tree(
+    current_revision: "DocumentWithRevisions",
+    doc: "DocumentWithRevisions",
+    seen: List[str],
+):
+    r"""Prints the revision tree of a child revision.
+
+    Args:
+        None.
+
+    Returns:
+        None.
+    """
+    tab = "  " * _get_level(doc)
+
+    if doc.children:
+        if current_revision.revision_id == doc.revision_id:
+            print(tab + "└──>", end="")
+        else:
+            print(tab + "└──", end="")
+        print(doc.revision_id)
+        for each in doc.children:
+            seen.append(each.revision_id)
+            _print_child_tree(current_revision, each, seen)
+    else:
+        if current_revision.revision_id == doc.revision_id:
+            print(tab + "└──>", end="")
+        else:
+            print(tab + "└──", end="")
+        print(doc.revision_id)
+
+
+@dataclasses.dataclass
+class DocumentWithRevisions:
```

Comment:
Have we considered having this as a subclass of wrapper Document?  Another seemingly reasonable alternative is to upgrade the existing wrapper Document class so that it knows how to handle revisions to.  (Maybe these were already discussed and rejected in the design and/or prototype, please feel free to just point me to the design doc.)

Reply:
I agree, I'm still not sure I understand why this extra class is needed. Or how this fits into the larger scheme of the library.

Reply:
https://github.com/googleapis/python-documentai-toolbox/pull/128#discussion_r1254734536

the discussion about which option was the preferred one was done in design review meeting and not in the design doc. I don't know if making it a subclass of Document would work out that well since the two are completely different, it might be better to have a meeting about this but in short the Document wrapper class reads json files and the logic it uses to construct the wrapped Document is different from how it is constructed in the Document with Revisions. Document with revisions reads `dp.bp` files which are binary protobuf messages from my understanding and the logic that is used to construct a Document is different from the normal sharding process since there are base_document, revisions, and revision history that is not present in normal `BatchProcessDocument` or `ProcessDocument` responses